katdal/sensordata.py

```diff
@@ -126,34 +186,112 @@ class TelstateSensorData(SensorData):
+            # This forces sensor values to be 1-D at all times (an invariant).
+            self.dtype = test_data.dtype if test_data.ndim == 1 else np.object
+
+    def __getitem__(self, key):
+        """Extract timestamp and value of each sensor data point."""
+        if key == 'timestamp':
+            return np.asarray(self._data[key])
+        elif key == 'value':
+            # Unpack everything first, otherwise old files will be a mess
+            values = [_h5_telstate_unpack(s) for s in self._data[key]]
+            if self.dtype == np.object:
+                values = [HashableValueWrapper.wrap(value) for value in values]
+            values = np.array(values)
+            # Rediscover dtype from values (mostly to fix variable-length
+            # string types now that we have seen all the data)
+            self.dtype = values.dtype
+            return values
+        else:
+            raise ValueError("Sensor %r data has no key '%s'" % (self.name, key))
+
+
+class TelstateSensorData(SensorData):
+    """Raw (uninterpolated) sensor data stored in original TelescopeState.
+
+    This wraps sensor data stored in a TelescopeState object. The data is
+    only read out on item access and then cached on the object. The caching
+    should be fine as a SensorData object is typically replaced by either
+    a CategoricalData object or a NumPy array as part of sensor extraction,
+    right after the caching occurs.
+
+    Parameters
     ----------
-    dtype : :class:`numpy.dtype` object
-        Type of sensor data, as a NumPy dtype
+    telstate : :class:`katsdptelstate.TelescopeState` object
+        Telescope state object
+    name : string
+        Sensor name, also used as telstate key
+
+    Raises
+    ------
+    KeyError
+        If sensor name is not found in telstate or it is an attribute instead
 
     """
-    def __init__(self, data, name=None):
-        super(TelstateSensorData, self).__init__(data, name)
-        # Unpickle first value to derive dtype (should be simple for sensors)
-        self.dtype = np.array([_telstate_unpack(data['value'][0])]).dtype
+    def __init__(self, telstate, name):
+        self._telstate = telstate
+        # This cache simplifies separate 'timestamp' / 'value' access pattern
+        self._value_times = []
+        if name not in telstate:
+            raise KeyError('No sensor named %r in telstate (key not found)' %
+                           (name,))
+        if telstate.is_immutable(name):
+            raise KeyError("No sensor named %r in telstate (it's an attribute)" %
+                           (name,))
+        test_data = np.array([telstate[name]])
+        # Beware array-valued sensors; treat their values as opaque objects
+        # otherwise they will turn 'value' into a mega-array of e.g. float.
+        # This forces sensor values to be 1-D at all times (an invariant).
+        dtype = test_data.dtype if test_data.ndim == 1 else np.object
```

Comment:
This logic (and the comments) seem to be repeated in a few places. Perhaps it should be put into a helper function? It also has the potential to be optimised later in a way that doesn't require constituting the full array.